db/selector.go

```diff
@@ -0,0 +1,145 @@
+	"context"
+	"database/sql"
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+type typeSelector[T any] struct {
+	wrapped MappedExecutor
+}
+
+// NewTypeSelector returns an object which can be used to automagically query
+// the provided type-mapped database for rows of the parameterized type.
+func NewTypeSelector[T any](wm MappedExecutor) *typeSelector[T] {
+	return &typeSelector[T]{wrapped: wm}
+}
+
+// SelectRows combines the best features of gorp, the go stdlib, and generics.
+// It uses the type parameter of the typeSelector object to automatically look
+// up the proper table name and columns to select. It returns an iterable which
+// yields fully-populated objects of the parameterized type directly. The given
+// clauses MUST be only the bits of a sql query from "WHERE ..." onwards; if
+// they contain any of the "SELECT ... FROM ..." portion of the query it will
+// result in an error. The caller is responsible for calling `Rows.Close()`
+// when they are done with the query.
+func (ts typeSelector[T]) SelectRows(ctx context.Context, clauses string, args ...interface{}) (*typeRows[T], error) {
+	// Look up the table to use based on the type of this TypeSelector.
+	var t T
+	tableMap, err := ts.wrapped.TableFor(reflect.TypeOf(t), false)
+	if err != nil {
+		return nil, fmt.Errorf("database model type not mapped to table name: %w", err)
+	}
+
+	return ts.SelectRowsFrom(ctx, tableMap.TableName, clauses, args...)
+}
+
+// SelectRowsFrom is the same as SelectRows, but it additionally takes a table
+// name to select from, rather than automatically computing the table name from
+// gorp's DbMap.
+func (ts typeSelector[T]) SelectRowsFrom(ctx context.Context, tablename string, clauses string, args ...interface{}) (*typeRows[T], error) {
+	// Look up the table to use based on the type of this TypeSelector.
+	var throwaway T
+	t := reflect.TypeOf(throwaway)
+	tableMap, err := ts.wrapped.TableFor(t, false)
+	if err != nil {
+		return nil, fmt.Errorf("database model type not mapped to table name: %w", err)
+	}
+
+	// Extract the list of column names from the type's struct tags.
+	var columns []string
+	for _, column := range tableMap.Columns {
+		columns = append(columns, column.ColumnName)
+	}
+
+	// Iterate over the columns in the order they appear. For each, find the field
+	// on the struct type that has a matching `db:"colname"` struct tag. Save that
+	// field's index into a map for quick lookup later.
+	colIndexToFieldIndex := make([][]int, len(columns))
+	for i, column := range columns {
+		structField, found := t.FieldByNameFunc(func(fieldName string) bool {
+			structField, _ := t.FieldByName(fieldName)
+			tagColumn := structField.Tag.Get("db")
+			return tagColumn == column
+		})
+		if !found {
+			// This should never happen, as the columns were derived from the struct
+			// fields in the first place.
+			return nil, fmt.Errorf("no struct field with tag matching column %q", column)
+		}
+		colIndexToFieldIndex[i] = structField.Index
+	}
```

Comment:
Given that you're already looking at the struct tag, looking up the `TableFor` and getting the columns seems a little redundant, right?

What about this: the list of columns in the query will always be in the same order of definition in the struct (via [`reflect.VisibleFields`](https://pkg.go.dev/reflect#VisibleFields). So we go from `reflect.VisibleFields()` -> `.Tag.Get("db")` -> list of column names for query.

Then in `Get()`, we call `.Field(i)` to get the appropriate field; no idxMap needed.

Reply:
Talked this through offline. Conclusion: I'm going to go audit all of our current model types. If they all abide by these restrictions (no nested structs, all fields are tagged with db columns) then I'll go ahead and do this simplification now -- it's clear that we don't use those features from gorp so let's just cut them out of the loop right now. If some of those model types are more complex objects, then we'll land this as-is with just a comment noting that we could take a simpler path in the future.

I'll report my findings here shortly.

Reply:
I've gone ahead and simplified this. I think that, in the spirit of https://twitter.com/ejames_c/status/1539847144622145536, the direction we're headed with this is towards a fully-fledged but greatly-simplified ORM of our own. With that in mind, I think it is worth scrapping the bits of struct-to-table mapping that gorp does that we don't use (such as nested structs, complex struct tags, automatic type conversions, etc) and setting ourselves up for a very strict and streamlined interface.